llvm/lib/Target/AIE/AIEAddressSpaceLowering.cpp

```diff
@@ -0,0 +1,115 @@
+//===--------- AIEAddressSpaceLowering.cpp - Retype pointer to P0  --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseSubtarget.h"
+#include "llvm/CodeGen/GlobalISel/CSEInfo.h"
+#include "llvm/CodeGen/GlobalISel/CSEMIRBuilder.h"
+#include "llvm/CodeGen/GlobalISel/GenericMachineInstrs.h"
+#include "llvm/CodeGen/GlobalISel/MachineIRBuilder.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineModuleInfo.h"
+#include "llvm/CodeGen/TargetPassConfig.h"
+#include "llvm/InitializePasses.h"
+
+#define DEBUG_TYPE "aie-addrspace-lowering"
+
+using namespace llvm;
+
+static const char AIE_POINTER_RETYPE[] = "AIE Address Space Lowering Pass";
+
+namespace {
+
+class AIEAddressSpaceLowering : public MachineFunctionPass {
+public:
+  static char ID;
+  AIEAddressSpaceLowering() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override {
+    MachineRegisterInfo &MRI = MF.getRegInfo();
+    TargetPassConfig &TPC = getAnalysis<TargetPassConfig>();
+    // Enable CSE to set up GISelObserverWrapper object
+    GISelCSEAnalysisWrapper &Wrapper =
+        getAnalysis<GISelCSEAnalysisWrapperPass>().getCSEWrapper();
+    auto *CSEInfo = &Wrapper.get(TPC.getCSEConfig());
+    std::unique_ptr<MachineIRBuilder> Builder =
+        CSEInfo ? std::make_unique<CSEMIRBuilder>()
+                : std::make_unique<MachineIRBuilder>();
+    Builder->setMF(MF);
+    MachineIRBuilder &MIB = *Builder;
+    // Set Observer to inform change of register type or change in op-code
+    GISelObserverWrapper Observer;
+    if (CSEInfo) {
+      Observer.addObserver(CSEInfo);
+      MIB.setChangeObserver(Observer);
+    }
+
+    for (MachineBasicBlock &MBB : MF) {
+      for (MachineInstr &MI : MBB) {
+        processMI(MI, MRI, MIB, Observer);
+      }
+    }
+    return true;
```

Comment:
Here it is better to return true only if you change some instruction. In this way, it is better to have a bool return type for `processMI`.